File: driver-service/src/main/java/com/tompang/carpool/driver_service/dto/DriverRegistrationResponseDto.java
package com.tompang.carpool.driver_service.dto;

import java.time.LocalDateTime;

import com.tompang.carpool.driver_service.model.DriverRegistration;
import com.tompang.carpool.driver_service.model.RegistrationStatus;

import lombok.Builder;
import lombok.Data;

@Data
@Builder
public class DriverRegistrationResponseDto {
    public String id;
    public String userId;
    public String vehicleRegistrationNumber;
    public String vehicleMake;
    public String vehicleModel;
    public LocalDateTime createdAt;
    public RegistrationStatus status;
    public String rejectedReason; // null if not failed status
    public String signedImageUrl; // null if not provided

    public static DriverRegistrationResponseDto fromEntity(DriverRegistration driver) {
        DriverRegistrationResponseDto dto = DriverRegistrationResponseDto.builder()
            .id(driver.getId())
            .userId(driver.getUserId())
            .vehicleRegistrationNumber(driver.getVehicleRegistrationNumber())
            .vehicleMake(driver.getVehicleMake())
            .vehicleModel(driver.getVehicleModel())
            .createdAt(driver.getCreatedAt())
            .status(driver.getRegistrationStatus())
            .build();
        
        if (driver.getRegistrationStatus().equals(RegistrationStatus.FAILED)) {
            if (driver.getManualReview() != null && driver.getManualReview().getFailReason() != null) {
                dto.setRejectedReason(driver.getManualReview().getFailReason());
            } else {
                // automatically rejected
                dto.setRejectedReason("Automatic validation failed");
            }
        }

        return dto;
    }
}

File: webapp-bff/src/backend/driver/dto/driver-registration-response.dto.ts
import { Type } from "class-transformer";
import { IsDate, IsString } from "class-validator";

export type RegistrationStatus = "PENDING" | "PENDING_MANUAL_REVIEW" | "SUCCESS" | "FAILED" | "INACTIVE";

export class DriverRegistrationResponseDto {

  @IsString()
  id: string;
  
  @IsString()
  userId: string;
  
  @IsString()
  vehicleRegistrationNumber: string;
  
  @IsString()
  vehicleMake: string;
  
  @IsString()
  vehicleModel: string;

  @Type(() => Date)
  @IsDate()
  createdAt: Date;

  @IsString()
  status: RegistrationStatus;
  rejectedReason: string | null; // null if not failed status
  signedImageUrl: string | null; // null if not provided
}